src/Analysis/Ast/Impl/Analyzer/PythonAnalyzerSession.cs

```diff
@@ -230,135 +255,285 @@ internal sealed class PythonAnalyzerSession {
+            var entry = node.Value;
+            if (!CanUpdateAnalysis(entry, _walker.Version, out var module, out var ast)) {
+                return;
             }
+            var startTime = stopWatch.Elapsed;
+            AnalyzeEntry(node, entry, module, ast, _walker.Version);
+
+            LogCompleted(node, module, stopWatch, startTime);
         }
 
         private void AnalyzeEntry() {
             var stopWatch = _log != null ? Stopwatch.StartNew() : null;
             try {
-                if (!CanUpdateAnalysis(_entry, null, Version, out var module, out var ast, out var currentAnalysis)) {
+                if (!CanUpdateAnalysis(_entry, Version, out var module, out var ast)) {
                     return;
                 }
                 var startTime = stopWatch?.Elapsed ?? TimeSpan.Zero;
                 AnalyzeEntry(null, _entry, module, ast, Version);
 
-                LogCompleted(null, module, stopWatch, startTime);
+                LogCompleted(module, stopWatch, startTime);
             } catch (OperationCanceledException oce) {
                 _entry.TryCancel(oce, Version);
                 LogCanceled(_entry.Module);
             } catch (Exception exception) {
                 _entry.TrySetException(exception, Version);
-                LogException(_entry.Module, exception);
+                LogException(_entry, exception);
             } finally {
                 stopWatch?.Stop();
             }
         }
 
-        private bool CanUpdateAnalysis(
-            PythonAnalyzerEntry entry,
-            IDependencyChainNode<PythonAnalyzerEntry> node,
-            int version,
-            out IPythonModule module,
-            out PythonAst ast,
-            out IDocumentAnalysis currentAnalysis) {
+        private void AnalyzeLoop(IDependencyChainLoopNode<PythonAnalyzerEntry> loopNode, Stopwatch stopWatch) {
+            var version = _walker.Version;
+            var entries = new Dictionary<AnalysisModuleKey, (IPythonModule Module, PythonAnalyzerEntry Entry)>();
+            var variables = new Dictionary<(AnalysisModuleKey Module, string Name), int>();
+            var importNames = new List<(AnalysisModuleKey From, int FromPosition, AnalysisModuleKey To, string ToName)>();
+            var cachedVariables = new Dictionary<AnalysisModuleKey, IVariableCollection>();
+            var asts = new Dictionary<AnalysisModuleKey, PythonAst>();
+            var startTime = stopWatch.Elapsed;
+
+            // Note: loop analysis is not cancellable. The reason is that when smaller loop
+            // appears inside a larger loop gets canceled, it will not be re-walked during
+            // the outer loop analysis. For example, functools <=> _functools loop and
+            // the related CircularDependencyFunctools test.
+            foreach (var entry in loopNode.Values) {
+                if (!CanUpdateAnalysis(entry, Version, out var module, out var ast)) {
```

Comment:
Needs `ActivityTracker.OnEnqueueModule(entry.Module.FilePath);` here to match `Analyze()`, otherwise our perf telemetry is going to be wrong.

Reply:
I think I can make this change directly on github; I'd like to get this PR merged so we can start merging everything else for a release.

Reply:
I am not entire sure it is needed. Activity tracks what is actually enqueued, but in loops some bypass regular analysis.